Batteries/Tactic/Init.lean

```diff
@@ -31,23 +31,31 @@ elab (name := exacts) "exacts " "[" hs:term,* "]" : tactic => do
     evalTactic (← `(tactic| exact $stx))
   evalTactic (← `(tactic| done))
 
+/--
+`by_contra_core` is the component of `by_contra` that turns the goal into the form `p → False`.
+`by_contra h` is defined as `by_contra_core` followed by `rintro h`.
+* If the goal is a negation `¬q`, the goal becomes `q → False`.
+* If the goal has a `Decidable` instance, it uses `Decidable.byContradiction` instead of
+  `Classical.byContradiction`.
+-/
+scoped macro "by_contra_core" : tactic => `(tactic| first
+  | guard_target = Not _; change _ → False
+  | refine @Decidable.byContradiction _ _ ?_
+  | refine @Classical.byContradiction _ ?_)
+
 /--
 `by_contra h` proves `⊢ p` by contradiction,
 introducing a hypothesis `h : ¬p` and proving `False`.
 * If `p` is a negation `¬q`, `h : q` will be introduced instead of `¬¬q`.
 * If `p` is decidable, it uses `Decidable.byContradiction` instead of `Classical.byContradiction`.
-* If `h` is omitted, the introduced variable `_: ¬p` will be anonymous.
+* If `h` is omitted, the introduced variable will be called `this`.
 -/
-macro (name := byContra) tk:"by_contra" e?:(ppSpace colGt binderIdent)? : tactic => do
-  let e := match e? with
-    | some e => match e with
-      | `(binderIdent| $e:ident) => e
-      | e => Unhygienic.run `(_%$e) -- HACK: hover fails without Unhygienic here
-    | none => Unhygienic.run `(_%$tk)
```

Comment:
@fgdorais 

> I was surprised to learn that `by_contra` without arguments had hover information with the type of the introduced variable. I found this surprising and confusing, but I also couldn't figure out how to keep this feature with the new implementation, so this feature has been removed.

The way this works is that the above line transfers the span of `tk` which is the `by_contra` token to the `_` which is then elaborated and acquires the type hover. To replicate this in the new implementation you should replace ``mkIdent `this`` by ``mkIdentFrom tk `this`` where `tk` is bound by `by_contra%$tk` in the macro_rules pattern.